File: problemas_beecrowd/b2447.cpp
// https://www.beecrowd.com.br/judge/pt/problems/view/2447

#include <bits/stdc++.h>

using namespace std;

#define _ std::ios::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);
#define dbg(x) cout << #x << " = " << x << endl
#define endl "\n"
#define MOD 1'000'000'007

typedef unsigned long long ull;
typedef long long ll;

ull arrangement(ull a, ull b) {
    ull r {1};
    if (a < b) return 0;
    for (ull i {a}; i > a - b; i--)
        r = (r * i) % MOD;
    return r;
}

ull perm(ull a) {
    return arrangement(a, a);
}

ull pow2(ull num) {
    return 1 << num;
}

ull compute_wining_positions(ull n, ull g) {
    if (g) return 0;
    return perm(1 << n);
}

ull additive_inverse(ll num, ull mod) {
    return (mod - abs(num) % mod) % mod;
}

ull compute_positions_that_lose_at_k(ull k, ull n, ull l, ull g) {
    if (k == n + 1)
        return compute_wining_positions(n, g);

    ull key = pow2(k - 1);

    if (l < key - 1 or not g)
        return 0;

    return perm(pow2(n) - pow2(k)) * arrangement(l, key - 1) % MOD * pow2(n) % MOD *
            ((arrangement(g + l - key + 1, key) + additive_inverse(-arrangement(l - key + 1, key), MOD)) % MOD) % MOD;
}

int main() {
    ull n, k; cin >> n >> k;
    ull juca; cin >> juca;
    ull lower {0}, greater{0};
    for (ull i {0}, card_players = (1 << n) - 1; i < card_players; i++) {
        ull player; cin >> player;
        if (player < juca) lower++;
        else greater++;
    }

    cout << compute_positions_that_lose_at_k(k, n, lower, greater) << endl;

    return 0;
}

File: problemas_beecrowd/b2447.py
MOD = 1_000_000_007


def arrangement(a: int, b: int) -> int:
    if a < b:
        return 0
    r = 1
    for i in range(a - b + 1, a + 1):
        r = (r * i) % MOD
    return r


def perm(a: int) -> int:
    return arrangement(a, a)


def pow2(num: int) -> int:
    return (1 << num) % MOD


def fatorial_em_intervalo(a, b=0) -> int:
    x = 1

    for i in range(a, b, -1):
        x *= i
        x = x % MOD

    return x


def arranjo(n, k) -> int:
    if k > n:
        return 0

    return fatorial_em_intervalo(n, n - k)


def compute_wining_positions(n: int, g: int) -> int:
    if g:
        return 0
    return perm(pow2(n))


def compute_positions_that_lose_at_k(k: int, n: int, l: int, g: int) -> int:
    if k == n + 1:
        return compute_wining_positions(n, g)
    if l < pow2(k - 1) - 1 or not g:
        return 0

    key = pow2(k - 1)

    x = (arranjo(l, key - 1) * 2 ** n
         * (arranjo(g + l - key + 1, key) - arranjo(l - key + 1, key))
         * fatorial_em_intervalo(2 ** n - 2 ** k))
    print(x % MOD)

    return (arrangement(l, key - 1) * pow2(n) % MOD *
            (arrangement(g + l - key + 1, key) - arrangement(l - key + 1, key)) % MOD *
            perm(pow2(n) - pow2(k)) % MOD)


def main() -> None:
    n, k = map(int, input().split())
    j = int(input())
    l, g = 0, 0
    for i in range(pow2(n) - 1):
        p = int(input())
        if p < j:
            l += 1
        else:
            g += 1
    print(compute_positions_that_lose_at_k(k, n, l, g))


if __name__ == '__main__':
    main()
